devito/types/basic.py

```diff
@@ -1147,13 +1152,16 @@ def _evaluate(self, **kwargs):
             return self
 
         # Base function
-        retval = self.function
+        retval = 1 / self.function if self._avg_mode == 'harmonic' else self.function
         # Apply interpolation from inner most dim
         for d, i in self._grid_map.items():
-            retval = retval.diff(d, 0, fd_order=2, x0={d: i})
+            retval = retval.diff(d, deriv_order=0, fd_order=2, x0={d: i}).evaluate
+        if self._avg_mode == 'harmonic':
+            retval = 1 / retval
+
         # Evaluate. Since we used `self.function` it will be on the grid when evaluate
         # is called again within FD
-        return retval.evaluate
+        return retval.evaluate.expand()
```

Comment:
why the need to expand ?

also, not that we're calling `.evaluate` multiple times here -- above in the `for d, i in self._grid_map` loop at each iteration, and then again here.

THere's something weird about this

Reply:
Without exapnd you get horrible stuff like 05*(0.5*(0.5 ... instead of 0.125.

This is a trivial square/cube averaging that should be a plain expression anything else is just messing up with the compiler for no reason